pyomo/core/expr/expr_pyomo5.py

```diff
@@ -0,0 +1,3434 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright 2017 National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from __future__ import division
+
+_using_chained_inequality = True
+
+#
+# These symbols are part of pyomo.core.expr
+#
+__public__ = ['linear_expression', 'nonlinear_expression', 'inequality']
+#
+# These symbols are part of pyomo.core.expr.current
+#
+__all__ = (
+'linear_expression',
+'nonlinear_expression',
+'inequality',
+'mutable_sum_context',
+'mutable_linear_context',
+'decompose_term',
+'clone_counter',
+'clone_counter_context',
+'clone_expression',
+'evaluate_expression',
+'identify_components',
+'identify_variables',
+'identify_mutable_parameters',
+'expression_to_string',
+'ExpressionBase',
+'EqualityExpression',
+'RangedExpression',
+'InequalityExpression',
+'ProductExpression',
+'MonomialTermExpression',
+'PowExpression',
+'ExternalFunctionExpression',
+'GetItemExpression',
+'Expr_ifExpression',
+'LinearExpression',
+'ReciprocalExpression',
+'NegationExpression',
+'SumExpression',
+'UnaryFunctionExpression',
+'AbsExpression',
+'compress_expression',
+'NPV_NegationExpression',
+'NPV_ExternalFunctionExpression',
+'NPV_PowExpression',
+'NPV_ProductExpression',
+'NPV_ReciprocalExpression',
+'NPV_SumExpression',
+'NPV_UnaryFunctionExpression',
+'NPV_AbsExpression',
+'SimpleExpressionVisitor',
+'ExpressionValueVisitor',
+'ExpressionReplacementVisitor',
+'LinearDecompositionError',
+'SumExpressionBase',
+'_MutableSumExpression',    # This should not be referenced, except perhaps while testing code
+'_MutableLinearExpression',     # This should not be referenced, except perhaps while testing code
+'_decompose_linear_terms',      # This should not be referenced, except perhaps while testing code
+'_chainedInequality',           # This should not be referenced, except perhaps while testing code
+'_using_chained_inequality',           # This should not be referenced, except perhaps while testing code
+'_generate_sum_expression',                 # Only used within pyomo.core.expr
+'_generate_mul_expression',                 # Only used within pyomo.core.expr
+'_generate_other_expression',               # Only used within pyomo.core.expr
+'_generate_intrinsic_function_expression',  # Only used within pyomo.core.expr
+'_generate_relational_expression',          # Only used within pyomo.core.expr
```

Comment:
Why are "`_`" methods in `__all__`?

Reply:
As the comments note, these shouldn't be exposed.  But I needed to expose them for testing.  And there were some instances where we were using the this functionality unexpectedly.  Also, this includes deprecated functionality.